### ISSUE TITLE
Reuse forward pass from `trace` in backprop GFI methods

### ISSUE THREAD

**Issue by alex-lew** [probcomp/GenPyTorch.jl#2]
Currently, when the user runs `simulate` or `generate`, a Torch generative function does not record operations for AD -- we run with `torch.no_grad()`. This is good: it is faster and more memory-efficient. But if the user generates a trace to later perform `choice_gradients` or `accumulate_param_gradients!` on, it is somewhat wasteful: a forward pass was used to generate the trace, but we need to run the forward pass again in order to record the necessary metadata for gradients.

One possibility would be to allow users to set a flag on the generative function (i.e., the GF itself could be mutable), controlling whether forward-mode executions recorded info for gradients. Then, when a gradient operation is required on a trace, we could check to see whether the necessary information was available in the trace, and if so, avoid a second forward pass.
